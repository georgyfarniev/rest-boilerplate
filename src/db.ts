import mongoose, { Document, Schema } from 'mongoose'

//#region models

const enumValues = (enumeration: any) => Object
  .keys(enumeration)
  .reduce((p: string[], c: string) => ([...p, enumeration[c]]), [])

enum PetType {
  Dog = 'dog',
  Cat = 'cat',
  Squirrel = 'squirrel'
}

interface IPet extends Document {
  name: string
  type: PetType
}

const PetSchema: Schema = new Schema({
  name: String,
  type: { type: String, enum: enumValues(PetType) }
})

const Pet = mongoose.model<IPet>('Pet', PetSchema)

//#endregion

async function connect(url: string) {
  console.log('connecting to db')

  mongoose.set('useUnifiedTopology', true)
  mongoose.set('useNewUrlParser', true)

  await mongoose.connect(url)

  console.log('db connected')

  return mongoose;
}

export default {
  connect,
  Pet
}