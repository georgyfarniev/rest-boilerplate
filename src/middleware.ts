import { Context, Next } from 'koa'
import * as yup from 'yup'

export function validateBody(schema: yup.ObjectSchema) {
  return async (ctx: Context, next: Next) => {
    try {
      await schema.validate(ctx.request.body)
      return next()
    } catch (err) {
      ctx.status = 400
      ctx.body = err
    }
  }
}
