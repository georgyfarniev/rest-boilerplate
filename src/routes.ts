import Router from '@koa/router'
import { Context } from 'koa'
import db from './db'

const router = new Router()

// List pets
router.get('/pets', async (ctx: Context) => {
  ctx.body = await db.Pet.find().lean()
})


export default router.routes()
