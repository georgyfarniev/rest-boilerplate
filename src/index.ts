import Koa from 'koa'
import bodyParser from 'koa-bodyparser-ts';
import Router from '@koa/router'
import cors from '@koa/cors';
import routes from './routes'
import db from './db'
import { Server } from 'http'

interface IStartServerOpts {
  port: number
  dbUrl: string
}

// port and db are parameterized for testing
export async function startServer({ port, dbUrl }: IStartServerOpts) {
  const dbc = await db.connect(dbUrl)

  const app = new Koa()

  app.use(cors())
  app.use(bodyParser())

  const router = new Router();

  router.prefix('/api').use('/routes', routes)

  app.use(router.routes())

  const server: Server = await new Promise((resolve) => {
    const server = app.listen(port, () => resolve(server))
  })

  return async () => Promise.all([ dbc.disconnect(), server.close() ])
}

const PORT = 1337;
const DB_URL = 'mongodb://localhost:27017/rest-boilerplate'

async function main() {
  await startServer({ port: PORT, dbUrl: DB_URL})
  console.log('started app on:', PORT)
}

if (require.main === module) {
  main().catch(console.error)
}
