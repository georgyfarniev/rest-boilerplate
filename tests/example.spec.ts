import 'jest'
import axios from 'axios'
import { startServer } from '../src/index'
import db from '../src/db'

// TODO: get port automatically
const TEST_PORT = 10101

const api = axios.create({
  baseURL: `http://localhost:${TEST_PORT}/api/routes`
})

describe('rest api test example', () => {
  let stopServer: () => void

  beforeAll(async () => {
    stopServer = await startServer({ dbUrl: process.env.MONGO_URL, port: TEST_PORT })
  })

  afterAll(async () => stopServer())

  test('list pets when empty', async () => {
    await expect(api.get('pets')).resolves.toHaveProperty('data', [])
  })

  test('test one pet exists', async () => {
    await db.Pet.create({ name: 'Manis', type: db.PetType.Cat })

    const { data } = await api.get('pets')

    expect(data).toHaveLength(1)
    expect(data[0]).toMatchSnapshot({
      _id: expect.any(String)
    })
  })
})
