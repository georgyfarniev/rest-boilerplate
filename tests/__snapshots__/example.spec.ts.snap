// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest api test example test one pet exists 1`] = `
Object {
  "__v": 0,
  "_id": Any<String>,
  "name": "Manis",
  "type": "cat",
}
`;
